### PR TITLE
Workers Followup + cOctopusWorkerPool

### PR DESCRIPTION
**Pull request by stopthatastronaut**
Adds a Worker Pool resource, tidies up some minor issues from the cTentacleAgent WorkerPool PR. Fixes some stuff for PSScriptAnalyzer vLatest.

### CODE REVIEW

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 [hunk 1]
```diff
@@ -0,0 +1,248 @@
+# Workers and Worker Pools appeared in 2018.7.0
+# Scoping to Spaces in 2019.1.0. TODO: version checking!
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: still got a TODO

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        [PSCredential]$OctopusApiKey = [PSCredential]::Empty
+    )
+
+    $currentResource = Get-TargetResource -Url $Url `
+        -Ensure $Ensure `
+        -WorkerPoolName $WorkerPoolName `
+        -OctopusCredentials $OctopusCredentials `
+        -OctopusApiKey $OctopusApiKey
+
+    if ($Ensure -eq "Absent" -and $currentResource.Ensure -eq "Present") {
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: Drift in the `WorkerPoolDescription` field is not handled

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            -OctopusApiKey $OctopusApiKey
+    }
+    elseif ($Ensure -eq "Present" -and $currentResource.Ensure -eq "Absent") {
+        New-WorkerPool -Url $Url `
+            -WorkerPoolName $WorkerPoolName `
+            -WorkerPoolDescription $WorkerPoolDescription `
+            -SpaceId $SpaceId `
+            -OctopusCredentials $OctopusCredentials `
+            -OctopusApiKey $OctopusApiKey
+
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: whitespace

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    }
+}
+
+function Test-TargetResource {
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSDSCUseVerboseMessageInDSCResource", "")]
+    [OutputType([boolean])]
+    param (
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string]$Url,
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: Does `Url` match what we've used in other files?

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+            -Ensure $Ensure `
+            -WorkerPoolName $WorkerPoolName `
+            -WorkerPoolDescription $WorkerPoolDescription `
+            -SpaceID $SpaceId `
+            -OctopusCredentials $OctopusCredentials `
+            -OctopusApiKey $OctopusApiKey)
+
+    $params = Get-ODSCParameter $MyInvocation.MyCommand.Parameters
+
+    $currentConfigurationMatchesRequestedConfiguration = $true
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: Can we use a common implementation of this in the helper file?

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        [string]$Url,
+        [string]$WorkerPoolName,
+        [PSCredential]$OctopusCredentials = [PSCredential]::Empty,
+        [PSCredential]$OctopusApiKey = [PSCredential]::Empty
+    )
+
+    $repository = Get-OctopusClientRepository -Url $Url `
+        -OctopusCredentials $OctopusCredentials `
+        -OctopusApiKey $OctopusApiKey
+
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: whitespace

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    $workerpool = $repository.WorkerPools.FindByName($WorkerPoolName)
+    return $workerpool
+}
+
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: whitespace

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+function Get-OctopusClientRepository
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: Can we move this to a common implementation?

File: OctopusDSC/DSCResources/cTentacleAgent/cTentacleAgent.psm1 [hunk 9]
```diff
@@ -441,7 +380,9 @@ function Set-TargetResource {
         Write-Verbose "Upgrading Tentacle..."
         $serviceName = (Get-TentacleServiceName $Name)
         Stop-Service -Name $serviceName
-        Install-Tentacle $tentacleDownloadUrl $tentacleDownloadUrl64 $TentacleHomeDirectory
+        Install-Tentacle    -tentacleDownloadUrl $tentacleDownloadUrl `
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cTentacleAgent/cTentacleAgent.psm1]: too many spaces

File: OctopusDSC/DSCResources/cTentacleAgent/cTentacleAgent.psm1 [hunk 10]
```diff
@@ -781,7 +720,9 @@ function New-Tentacle {
     )
 
 
-    Install-Tentacle $tentacleDownloadUrl $tentacleDownloadUrl64 $tentacleHomeDirectory
+    Install-Tentacle    -tentacleDownloadUrl $tentacleDownloadUrl `
```
Comment (matt-richardson) [inline OctopusDSC/DSCResources/cTentacleAgent/cTentacleAgent.psm1]: whitespace police

File: OctopusDSC/Examples/cOctopusWorkerPool.ps1 [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+$password = Get-Content .\ExamplePassword.txt | ConvertTo-SecureString
+$creds = New-Object PSCredential "username", $password
+
+Configuration SampleConfig
+{
+    Import-DscResource -Module OctopusDSC
+
+    Node "localhost"
+    {
+        cOctopusWorkerPool "Create 'Development' Environment"
```
Comment (matt-richardson) [inline OctopusDSC/Examples/cOctopusWorkerPool.ps1]: wrong name

File: README-cTentacleAgent.md [hunk 12]
```diff
@@ -87,6 +87,7 @@ When `State` is `Started`, the resource will ensure that the Tentacle windows se
 | `TentacleDownloadUrl`         | `string`                                                   | `https://octopus.com/downloads/latest/OctopusTentacle`        | The url to use to download the Tentacle, when run on an `x86` architecture. |
 | `TentacleHomeDirectory`       | `string`                                                   | `$($env:SystemDrive)\Octopus`                                 | The location where Octopus should store its working files, logs, packages etc. |
 | `TentacleServiceCredential`   | `PSCredential`                                             |                                                               | Credentials of the account used to run the Tentacle Service
+| `WorkerPools`                 | `string[]`                                                 |                                                               | Worker Pool Name(s) in which this tentacle participates as a worker.
```
Comment (matt-richardson) [inline README-cTentacleAgent.md]: other fields we handle as a string as well, and split. Worth mentioning?

File: Tests/Scenarios/Tentacle_Scenario_01_Install.ps1 [hunk 13]
```diff
@@ -114,6 +114,32 @@ Configuration Tentacle_Scenario_01_Install
             OctopusServerThumbprint = $ServerThumbprint
         }
 
+        cTentacleAgent WorkerTentacle
+        {
+            Ensure = "Present";
+            State = "Started";
+
+            # Tentacle instance name. Leave it as 'Tentacle' unless you have more
```
Comment (matt-richardson) [inline Tests/Scenarios/Tentacle_Scenario_01_Install.ps1]: unnecessary comments

File: Tests/Scenarios/Tentacle_Scenario_01_Install.ps1 [hunk 14]
```diff
@@ -114,6 +114,32 @@ Configuration Tentacle_Scenario_01_Install
...
+            Environments = $Environments;
+            Roles = $Roles;
+
+            # Optional settings
+            ListenPort = 10937;
+            DefaultApplicationDirectory = "C:\Applications"
+            PublicHostNameConfiguration = "ComputerName"
+            TentacleHomeDirectory = "C:\Octopus\WorkerTentacleHome"
+
+            WorkerPools = "Default Worker Pool"
```
Comment (matt-richardson) [inline Tests/Scenarios/Tentacle_Scenario_01_Install.ps1]: Might be worth putting it into the new worker pool?

File: Tests/Spec/server_scenario_01_install_spec.rb [hunk 15]
```diff
@@ -60,6 +60,8 @@
   it { should exist }
 end
 
+#TODO: worker pool
```
Comment (matt-richardson) [inline Tests/Spec/server_scenario_01_install_spec.rb]: TODO :)

File: Tests/Spec/tentacle_scenario_01_install_spec.rb [hunk 16]
```diff
@@ -136,6 +136,40 @@
...
+
+describe octopus_deploy_tentacle(ENV['OctopusServerUrl'], ENV['OctopusApiKey'], "WorkerTentacle") do
+  it { should exist }
+  it { should be_registered_with_the_server }
+  it { should be_online }
+  it { should be_listening_tentacle }
+  it { should be_in_environment('The-Env') }
+  it { should have_role('Test-Tentacle') }
+  it { should have_display_name("WorkerTentacle")}
+  # TODO check pool membership
```
Comment (matt-richardson) [inline Tests/Spec/tentacle_scenario_01_install_spec.rb]: TODO

File: Tests/Spec/tentacle_scenario_01_install_spec.rb [hunk 17]
```diff
@@ -136,6 +136,40 @@
...
+end
+
+describe windows_registry_key('HKEY_LOCAL_MACHINE\Software\Octopus\Tentacle\WorkerTentacle') do
```
Comment (matt-richardson) [inline Tests/Spec/tentacle_scenario_01_install_spec.rb]: This wont exist will it? It's a new shiny tentacle, so will only have the file in c:\programdata, right?

File: build-hyperv.ps1 [hunk 18]
```diff
@@ -1,20 +1,45 @@
...
+  [switch]$SkipPester,
+  [switch]$ServerOnly,
+  [switch]$TentacleOnly
 )
 
 . Tests/powershell-helpers.ps1
 
 Start-Transcript .\vagrant-hyperv.log -Append
 
+# Clear the OctopusDSCTestMode Env Var
```
Comment (matt-richardson) [inline build-hyperv.ps1]: Consider moving all this new code to a function.

Also consider moving the opening brace onto the same line as the `if`, to make it more concise (controversial i know)

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 (hunk 4 above)
Comment (stopthatastronaut) [reply OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: Matches `cOctopusEnvironment` yeah. I wonder if we should consider a breaking change though. It's not ideal.

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 (hunk 5 above)
Comment (stopthatastronaut) [reply OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: We probably can now that multiple resources rely on it yeah. definitely on the list.

File: Tests/Scenarios/Tentacle_Scenario_01_Install.ps1 (hunk 14 above)
Comment (stopthatastronaut) [reply Tests/Scenarios/Tentacle_Scenario_01_Install.ps1]: I needed to get the resource working before doing that - it's on the list to do

File: OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1 (hunk 2 above)
Comment (stopthatastronaut) [reply OctopusDSC/DSCResources/cOctopusWorkerPool/cOctopusWorkerPool.psm1]: Indeed it isn't right now. Same for space membership.

File: Tests/Spec/tentacle_scenario_01_install_spec.rb (hunk 17 above)
Comment (stopthatastronaut) [reply Tests/Spec/tentacle_scenario_01_install_spec.rb]: Yep, I think you're correct

**Pull request merged**